File: Sources/ScreenPlus/CameraPath.swift
import Foundation

/// Camera + cursor state at one instant, in source pixels (top-left origin).
struct CameraState {
    var centerX: Double
    var centerY: Double
    var zoom: Double
    var cursorX: Double
    var cursorY: Double
}

/// A critically damped spring. Much nicer than easing curves because it handles targets
/// that change mid-animation (e.g. the cursor keeps moving while we zoom).
struct Spring {
    var value: Double
    var velocity: Double = 0
    var omega: Double  // stiffness as angular frequency; settle time ≈ 4.6 / omega seconds
    var damping: Double = 1  // 1 = critically damped (no overshoot)

    mutating func step(to target: Double, dt: Double) {
        let accel = omega * omega * (target - value) - 2 * damping * omega * velocity
        velocity += accel * dt  // semi-implicit Euler: stable at our small dt
        value += velocity * dt
    }
}

/// Precomputes the whole camera animation for a recording.
///
/// Because we render after recording, the path can "see the future": zooms start slightly
/// before the click that triggered them.
final class CameraPath {
    struct Segment {
        var start: Double
        var end: Double
        var focusX: Double
        var focusY: Double
    }

    private static let dt = 1.0 / 240.0
    /// Start zooming this long before the activity that triggered it.
    private static let lead = 0.35
    /// Zoom segments closer than this are merged, to avoid a quick out-and-in.
    private static let mergeGap = 0.8
    /// Mouse travel (in points) that counts as "activity".
    private static let movementThreshold = 12.0

    let width: Double
    let height: Double
    let segments: [Segment]
    private let states: [CameraState]
    private let clickTimes: [Double]

    init(session: RecordingSession, settings: RenderSettings, duration: Double) {
        let W = Double(session.width), H = Double(session.height)
        width = W
        height = H
        clickTimes = session.clicks.map(\.t).sorted()

        let cursor = session.cursor.sorted { $0.t < $1.t }
        func rawCursor(at t: Double) -> (x: Double, y: Double) {
            CameraPath.interpolate(cursor, at: t) ?? (W / 2, H / 2)
        }

        // 1. Activity timestamps: clicks, plus meaningful mouse movement.
        var activity: [(t: Double, x: Double, y: Double)] = session.clicks.map { ($0.t, $0.x, $0.y) }
        if !settings.clicksOnly, var anchor = cursor.first {
            let threshold = Self.movementThreshold * session.pixelsPerPoint
            for s in cursor.dropFirst() where hypot(s.x - anchor.x, s.y - anchor.y) > threshold {
                activity.append((s.t, s.x, s.y))
                anchor = s
            }
        }
        // Ignore the last second: that's usually the mouse heading to the Stop button.
        activity = activity.filter { $0.t >= 0 && $0.t < duration - 1.0 }.sorted { $0.t < $1.t }

        // 2. Merge activity into zoom segments.
        var segments: [Segment] = []
        for a in activity {
            let start = max(0, a.t - Self.lead)
            let end = a.t + settings.idleTimeout
            if let last = segments.last, start <= last.end + Self.mergeGap {
                segments[segments.count - 1].end = max(last.end, end)
            } else {
                segments.append(Segment(start: start, end: end, focusX: a.x, focusY: a.y))
            }
        }
        self.segments = segments

        // 3. Simulate springs over the whole timeline.
        let cursorOmega = 40 * pow(0.2, settings.cursorSmoothing)
        let start = rawCursor(at: 0)
        var zoomSpring = Spring(value: 0, omega: 6)  // animates log(zoom) so in/out feel symmetric
        var cx = Spring(value: W / 2, omega: 4.5)
        var cy = Spring(value: H / 2, omega: 4.5)
        var curX = Spring(value: start.x, omega: cursorOmega)
        var curY = Spring(value: start.y, omega: cursorOmega)

        var targetX = W / 2, targetY = H / 2
        var segIndex = 0
        var activeSegment: Int?
        let count = Int(duration / Self.dt) + 2
        var states: [CameraState] = []
        states.reserveCapacity(count)

        for i in 0..<count {
            let t = Double(i) * Self.dt
            let raw = rawCursor(at: t)
            curX.step(to: raw.x, dt: Self.dt)
            curY.step(to: raw.y, dt: Self.dt)

            while segIndex < segments.count, segments[segIndex].end < t { segIndex += 1 }
            let zoomed = segIndex < segments.count && segments[segIndex].start <= t
            let targetZoom = zoomed ? settings.zoomLevel : 1

            if zoomed {
                let viewW = W / targetZoom, viewH = H / targetZoom
                if activeSegment != segIndex {
                    activeSegment = segIndex
                    targetX = segments[segIndex].focusX
                    targetY = segments[segIndex].focusY
                }
                // Dead zone: only move the camera once the cursor nears the edge of the view.
                let halfX = viewW * 0.3, halfY = viewH * 0.3
                if curX.value > targetX + halfX { targetX = curX.value - halfX }
                if curX.value < targetX - halfX { targetX = curX.value + halfX }
                if curY.value > targetY + halfY { targetY = curY.value - halfY }
                if curY.value < targetY - halfY { targetY = curY.value + halfY }
                targetX = min(max(targetX, viewW / 2), W - viewW / 2)
                targetY = min(max(targetY, viewH / 2), H - viewH / 2)
            } else {
                activeSegment = nil
                targetX = W / 2
                targetY = H / 2
            }

            zoomSpring.step(to: log(targetZoom), dt: Self.dt)
            cx.step(to: targetX, dt: Self.dt)
            cy.step(to: targetY, dt: Self.dt)

            states.append(CameraState(centerX: cx.value, centerY: cy.value, zoom: exp(zoomSpring.value),
                                      cursorX: curX.value, cursorY: curY.value))
        }
        self.states = states
    }

    /// Camera state at time `t`, clamped so the view never leaves the recorded screen.
    func state(at t: Double) -> CameraState {
        let f = max(0, min(t / Self.dt, Double(states.count - 1)))
        let i = Int(f)
        let a = states[i], b = states[min(i + 1, states.count - 1)]
        let u = f - Double(i)
        func mix(_ x: Double, _ y: Double) -> Double { x + (y - x) * u }

        var s = CameraState(centerX: mix(a.centerX, b.centerX), centerY: mix(a.centerY, b.centerY),
                            zoom: max(1, mix(a.zoom, b.zoom)),
                            cursorX: mix(a.cursorX, b.cursorX), cursorY: mix(a.cursorY, b.cursorY))
        let viewW = width / s.zoom, viewH = height / s.zoom
        s.centerX = min(max(s.centerX, viewW / 2), width - viewW / 2)
        s.centerY = min(max(s.centerY, viewH / 2), height - viewH / 2)
        return s
    }

    /// 1 right at a click, decaying to 0 — used to "press" the cursor.
    func clickPulse(at t: Double) -> Double {
        let duration = 0.2
        guard let last = clickTimes.last(where: { $0 <= t }), t - last < duration else { return 0 }
        return 1 - (t - last) / duration
    }

    private static func interpolate(_ samples: [CursorSample], at t: Double) -> (x: Double, y: Double)? {
        guard let first = samples.first, let last = samples.last else { return nil }
        if t <= first.t { return (first.x, first.y) }
        if t >= last.t { return (last.x, last.y) }
        // Binary search for the first sample after t.
        var lo = 0, hi = samples.count - 1
        while lo < hi {
            let mid = (lo + hi) / 2
            if samples[mid].t <= t { lo = mid + 1 } else { hi = mid }
        }
        let a = samples[lo - 1], b = samples[lo]
        let u = b.t > a.t ? (t - a.t) / (b.t - a.t) : 1
        return (a.x + (b.x - a.x) * u, a.y + (b.y - a.y) * u)
    }
}

File: windows/src/ScreenPlus.Core/CameraPath.cs
namespace ScreenPlus;

/// <summary>Camera + cursor state at one instant, in source pixels (top-left origin).</summary>
public record struct CameraState(double CenterX, double CenterY, double Zoom, double CursorX, double CursorY);

/// <summary>
/// A critically damped spring. Much nicer than easing curves because it handles targets
/// that change mid-animation (e.g. the cursor keeps moving while we zoom).
/// </summary>
public struct Spring(double value, double omega)
{
    public double Value = value;
    public double Velocity = 0;
    /// <summary>Stiffness as angular frequency; settle time ≈ 4.6 / omega seconds.</summary>
    public double Omega = omega;
    /// <summary>1 = critically damped (no overshoot).</summary>
    public double Damping = 1;

    public void Step(double target, double dt)
    {
        var accel = Omega * Omega * (target - Value) - 2 * Damping * Omega * Velocity;
        Velocity += accel * dt;  // semi-implicit Euler: stable at our small dt
        Value += Velocity * dt;
    }
}

/// <summary>
/// Precomputes the whole camera animation for a recording.
///
/// Because we render after recording, the path can "see the future": zooms start slightly
/// before the click that triggered them.
/// </summary>
public sealed class CameraPath
{
    public readonly record struct Segment(double Start, double End, double FocusX, double FocusY);

    private const double Dt = 1.0 / 240.0;
    /// <summary>Start zooming this long before the activity that triggered it.</summary>
    private const double Lead = 0.35;
    /// <summary>Zoom segments closer than this are merged, to avoid a quick out-and-in.</summary>
    private const double MergeGap = 0.8;
    /// <summary>Mouse travel (in points) that counts as "activity".</summary>
    private const double MovementThreshold = 12.0;

    public double Width { get; }
    public double Height { get; }
    public IReadOnlyList<Segment> Segments { get; }

    private readonly CameraState[] _states;
    private readonly double[] _clickTimes;

    public CameraPath(RecordingSession session, RenderSettings settings, double duration)
    {
        double W = session.Width, H = session.Height;
        Width = W;
        Height = H;
        duration = double.IsFinite(duration) ? Math.Max(0, duration) : 0;
        _clickTimes = session.Clicks.Select(c => c.T).Order().ToArray();

        var cursor = session.Cursor.OrderBy(s => s.T).ToArray();
        (double x, double y) RawCursor(double t) => Interpolate(cursor, t) ?? (W / 2, H / 2);

        // 1. Activity timestamps: clicks, plus meaningful mouse movement.
        var activity = session.Clicks.Select(c => (t: c.T, x: c.X, y: c.Y)).ToList();
        if (!settings.ClicksOnly && cursor.Length > 0)
        {
            var anchor = cursor[0];
            var threshold = MovementThreshold * session.PixelsPerPoint;
            foreach (var s in cursor.Skip(1))
            {
                if (Hypot(s.X - anchor.X, s.Y - anchor.Y) > threshold)
                {
                    activity.Add((s.T, s.X, s.Y));
                    anchor = s;
                }
            }
        }
        // Ignore the last second: that's usually the mouse heading to the Stop button.
        activity = activity.Where(a => a.t >= 0 && a.t < duration - 1.0).OrderBy(a => a.t).ToList();

        // 2. Merge activity into zoom segments.
        var segments = new List<Segment>();
        foreach (var a in activity)
        {
            var start = Math.Max(0, a.t - Lead);
            var end = a.t + settings.IdleTimeout;
            if (segments.Count > 0 && start <= segments[^1].End + MergeGap)
                segments[^1] = segments[^1] with { End = Math.Max(segments[^1].End, end) };
            else
                segments.Add(new Segment(start, end, a.x, a.y));
        }
        Segments = segments;

        // 3. Simulate springs over the whole timeline.
        var cursorOmega = 40 * Math.Pow(0.2, settings.CursorSmoothing);
        var first = RawCursor(0);
        var zoomSpring = new Spring(0, 6);  // animates log(zoom) so in/out feel symmetric
        var cx = new Spring(W / 2, 4.5);
        var cy = new Spring(H / 2, 4.5);
        var curX = new Spring(first.x, cursorOmega);
        var curY = new Spring(first.y, cursorOmega);

        double targetX = W / 2, targetY = H / 2;
        var segIndex = 0;
        int? activeSegment = null;
        var count = (int)(duration / Dt) + 2;
        _states = new CameraState[count];

        for (var i = 0; i < count; i++)
        {
            var t = i * Dt;
            var raw = RawCursor(t);
            curX.Step(raw.x, Dt);
            curY.Step(raw.y, Dt);

            while (segIndex < segments.Count && segments[segIndex].End < t) segIndex++;
            var zoomed = segIndex < segments.Count && segments[segIndex].Start <= t;
            var targetZoom = zoomed ? settings.ZoomLevel : 1;

            if (zoomed)
            {
                double viewW = W / targetZoom, viewH = H / targetZoom;
                if (activeSegment != segIndex)
                {
                    activeSegment = segIndex;
                    targetX = segments[segIndex].FocusX;
                    targetY = segments[segIndex].FocusY;
                }
                // Dead zone: only move the camera once the cursor nears the edge of the view.
                double halfX = viewW * 0.3, halfY = viewH * 0.3;
                if (curX.Value > targetX + halfX) targetX = curX.Value - halfX;
                if (curX.Value < targetX - halfX) targetX = curX.Value + halfX;
                if (curY.Value > targetY + halfY) targetY = curY.Value - halfY;
                if (curY.Value < targetY - halfY) targetY = curY.Value + halfY;
                targetX = Math.Min(Math.Max(targetX, viewW / 2), W - viewW / 2);
                targetY = Math.Min(Math.Max(targetY, viewH / 2), H - viewH / 2);
            }
            else
            {
                activeSegment = null;
                targetX = W / 2;
                targetY = H / 2;
            }

            zoomSpring.Step(Math.Log(targetZoom), Dt);
            cx.Step(targetX, Dt);
            cy.Step(targetY, Dt);

            _states[i] = new CameraState(cx.Value, cy.Value, Math.Exp(zoomSpring.Value), curX.Value, curY.Value);
        }
    }

    /// <summary>Camera state at time <paramref name="t"/>, clamped so the view never leaves the recorded screen.</summary>
    public CameraState StateAt(double t)
    {
        var f = double.IsFinite(t) ? Math.Max(0, Math.Min(t / Dt, _states.Length - 1)) : 0;
        var i = (int)f;
        var a = _states[i];
        var b = _states[Math.Min(i + 1, _states.Length - 1)];
        var u = f - i;
        double Mix(double x, double y) => x + (y - x) * u;

        var s = new CameraState(Mix(a.CenterX, b.CenterX), Mix(a.CenterY, b.CenterY),
                                Math.Max(1, Mix(a.Zoom, b.Zoom)),
                                Mix(a.CursorX, b.CursorX), Mix(a.CursorY, b.CursorY));
        double viewW = Width / s.Zoom, viewH = Height / s.Zoom;
        s.CenterX = Math.Min(Math.Max(s.CenterX, viewW / 2), Width - viewW / 2);
        s.CenterY = Math.Min(Math.Max(s.CenterY, viewH / 2), Height - viewH / 2);
        return s;
    }

    /// <summary>1 right at a click, decaying to 0 — used to "press" the cursor.</summary>
    public double ClickPulse(double t)
    {
        const double duration = 0.2;
        // Last click at or before t.
        int lo = 0, hi = _clickTimes.Length;
        while (lo < hi)
        {
            var mid = (lo + hi) / 2;
            if (_clickTimes[mid] <= t) lo = mid + 1; else hi = mid;
        }
        if (lo == 0) return 0;
        var last = _clickTimes[lo - 1];
        return t - last < duration ? 1 - (t - last) / duration : 0;
    }

    private static (double x, double y)? Interpolate(CursorSample[] samples, double t)
    {
        if (samples.Length == 0) return null;
        CursorSample first = samples[0], last = samples[^1];
        if (t <= first.T) return (first.X, first.Y);
        if (t >= last.T) return (last.X, last.Y);
        // Binary search for the first sample after t.
        int lo = 0, hi = samples.Length - 1;
        while (lo < hi)
        {
            var mid = (lo + hi) / 2;
            if (samples[mid].T <= t) lo = mid + 1; else hi = mid;
        }
        CursorSample a = samples[lo - 1], b = samples[lo];
        var u = b.T > a.T ? (t - a.T) / (b.T - a.T) : 1;
        return (a.X + (b.X - a.X) * u, a.Y + (b.Y - a.Y) * u);
    }

    private static double Hypot(double x, double y) => Math.Sqrt(x * x + y * y);
}
